src/pgwire/src/protocol.rs

```diff
@@ -1005,6 +1005,39 @@ where
         self.flush().await
     }
 
+    // Converts a RowsFuture to a stream while also checking for connection close.
+    async fn row_future_to_stream(&self, rows: RowsFuture) -> Result<RowBatchStream, io::Error> {
+        let closed = async {
+            loop {
+                tokio::time::sleep(Duration::from_secs(1)).await;
+
+                // We've been waiting for rows for a bit, and the client may have
+                // disconnected. Check whether the socket is no longer readable and error
+                // if so.
+                match self.conn.ready(Interest::READABLE).await {
+                    Ok(ready) => {
+                        if ready.is_read_closed() {
+                            return io::Error::new(io::ErrorKind::Other, "connection closed");
+                        }
+                    }
+                    Err(err) => return err,
+                }
+            }
+        };
+        // Do not include self.coord_client.canceled() here because cancel messages
+        // will propagate through the PeekResponse.
+        tokio::select! {
+            err = closed => {
+                Err(err)
+            },
+            rows = rows => {
+                let (tx, rx) = tokio::sync::mpsc::unbounded_channel();
```

Comment:
Can't be a oneshot because this has to be a `RowBatchStream` for `TAIL` support. There is maybe a refactor, but there is a thing that needs multiple row batches, where SELECT is a special single batch case.

Reply:
Definitely another PR if needed though, yes.